hs/src/Reach/Optimize.hs

```diff
@@ -933,10 +1044,12 @@ instance {-# OVERLAPPING #-} (Optimize a, Optimize b, Optimize c, Optimize d, Op
 
 instance Optimize ETail where
   opt = \case
-    ET_Com m k -> mkCom ET_Com <$> opt m <*> opt k
+    ET_Com m k -> do
+      preFloat m
+      float ET_Com =<< mkCom ET_Com <$> opt m <*> opt k
```

Comment:
You duplicate this pattern above on line 990 and below on line 1074. You should make a function and a comment that this is "locally bottom-up"